### PR TITLE
node rpc: validateresource

### PR DESCRIPTION
**Pull request by tynes**
This PR adds a new Node RPC method called `validateresource`. It is useful for users who are in the process of creating their resource JSON to use with `sendupdate`. This PR adds error messages in the `fromJSON` methods of each of the records found in `lib/dns/resource.js`.

The RPC accepts a single object with the `records` property, which must be an array of objects.
Eventually, it could also accept a `version` property if another Resource version is implemented.

The error messages note the invalid JSON property name and the record type. This is so that good error messages can be returned to users. The error messages could be made generic (not including the record type in the error message) but that would mean that the RPC method handler itself would have to hold logic similar to `Resource.fromJSON` and keep track of which `record` failed its `fromJSON`. I don't like this approach as much since any new resource type will require its `fromJSON` logic to be added to the RPC method handler but am willing to update this PR if people are strongly opinionated in that way.

Closes https://github.com/handshake-org/hsd/issues/139

### CODE REVIEW

File: lib/node/rpc.js [hunk 1]
```diff
@@ -241,6 +241,7 @@ class RPC extends RPCBase {
     this.add('grindname', this.grindName);
     this.add('sendrawclaim', this.sendRawClaim);
     this.add('sendrawairdrop', this.sendRawAirdrop);
+    this.add('verifyresource', this.verifyResource);
```
Comment (pinheadmz) [inline lib/node/rpc.js:244]: Suggestion `validateresource`

**Comment by pinheadmz** [comment]
Looks pretty good, playing with the branch locally and might add a test.

Do you think some of these error message could be even more specific?
e.g.
`TXT record "txt" must be an array`
`Resource "records" must be an array`
`Resource "records" items must be objects`
`DS record keyType must be int < 0xffff`
`DS record "digest" must be string`
etc

File: lib/dns/resource.js [hunk 2]
```diff
@@ -615,10 +619,10 @@ class GLUE6 extends Struct {
   }
 
   fromJSON(json) {
-    assert(json && typeof json === 'object');
-    assert(json.type === 'GLUE6');
-    assert(isName(json.ns));
-    assert(IP.isIPv6String(json.address));
+    assert(json && typeof json === 'object', 'Invalid JSON in GLUE6 record.');
```
Comment (boymanjor) [inline lib/dns/resource.js:622]: nit: s/JSON/json/

**Comment by boymanjor** [comment]
> Looks pretty good, playing with the branch locally and might add a test.
> 
> Do you think some of these error message could be even more specific?
> e.g.
> `TXT record "txt" must be an array`
> `Resource "records" must be an array`
> `Resource "records" items must be objects`
> `DS record keyType must be int < 0xffff`
> `DS record "digest" must be string`
> etc

How about `Invalid DS record. Property "json" must be an Array.`

**Comment by tynes** [comment]
New commits pushed up, I followed similar error messaging to @boymanjor's suggestion here:
`DS record "digest" must be string`

**Comment by tynes** [comment]
This is ready for another round of review. I wrote tests for both success and failure cases.

File: lib/dns/resource.js [hunk 3]
```diff
@@ -437,13 +437,19 @@ class DS extends Struct {
...
-    assert((json.keyTag & 0xffff) === json.keyTag);
-    assert((json.algorithm & 0xff) === json.algorithm);
-    assert((json.digestType & 0xff) === json.digestType);
-    assert(typeof json.digest === 'string');
-    assert((json.digest.length >>> 1) <= 255);
+    assert(json && typeof json === 'object', 'Invalid DS record.');
+    assert(json.type === 'DS',
+      'Invalid DS record. Property "type" must be "DS".');
+    assert((json.keyTag & 0xffff) === json.keyTag,
+      'Invalid DS record. Property "keyTag" must be a u16.');
```
Comment (boymanjor) [inline lib/dns/resource.js:444]: nit: let's use `uint16` instead of `u16` to match prior art in the code base.

File: lib/node/rpc.js [hunk 4]
```diff
@@ -2477,6 +2478,27 @@ class RPC extends RPCBase {
     return proof.hash().toString('hex');
   }
 
+  async validateResource(args, help) {
+    if (help || args.length < 1 || args.length > 2)
+      throw new RPCError(errs.MISC_ERROR, 'validateresource'
+        + '"{"records": [{...}]}"');
```
Comment (pinheadmz) [inline lib/node/rpc.js:2484]: nit: this doesn't print out correctly.
Current:
```
$ hsd-rpc --network=regtest validateresource
validateresource"{"records": [{...}]}"
```

Suggestion:
```
$ hsd-rpc --network=regtest validateresource
validateresource '{"records": [{...}]}'
```

Note the single quotes as well, you may have to use `\'` in the code.

File: test/node-rpc-test.js [hunk 5]
```diff
@@ -130,5 +130,99 @@ describe('RPC', function() {
       const info = await nclient.execute('getblock', [hash]);
       assert.deepEqual(info.confirmations, -1);
     });
+
+    it('should validateresource (valid)', async () => {
+      const records = [
+        [{type: 'NS', ns: 'ns1.handshake.org.'}],
+        [{type: 'DS', keyTag: 0xffff, algorithm: 0xff, digestType: 0xff, digest: '00'.repeat(32)}],
+        [{type: 'TXT', txt: ['i like turtles', 'then who phone']}],
```
Comment (pinheadmz) [inline test/node-rpc-test.js:138]: lol wut.

**Comment by pinheadmz** [comment]
Great work on this dude, thanks for the thorough test. Only thing left for me is the help message.

File: lib/node/rpc.js (hunk 4 above)
Comment (tynes) [reply lib/node/rpc.js]: Good find! Pushed up a fix.

**Comment by pinheadmz** [comment]
ACK @  d5d98219eb3c8f24bfb948a3fd9dc51b7b1fca74

**Pull request merged**